### PR TITLE
AirfRANS: intermediate learning rate lr=5e-4

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

To bracket the optimal LR for AirfRANS we need three data points: 3e-4 (violet, default), 5e-4 (this run), and 1e-3 (gilbert). The intermediate point at 5e-4 will give us a finer picture of the LR sensitivity curve and help identify whether there's a sweet spot between 3e-4 and 1e-3.

## Instructions

```bash
cd target/icml2026
python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --epochs 150 \
  --lr 5e-4 \
  --optimizer lion \
  --batch-size 2 \
  --model senpai_transolver \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --model-slices 96 \
  --model-mlp-ratio 4 \
  --model-dropout 0.0 \
  --cosine-t-max 150 \
  --use-ema \
  --ema-decay 0.999 \
  --ema-start-step 50 \
  --use-lookahead \
  --surface-refine \
  --surface-refine-hidden 128 \
  --surface-refine-layers 2 \
  --weight-decay 1e-4 \
  --wandb-name "senku/airfrans-lr-5e4" \
  --wandb-group "radford-airfrans-round1"
```

## Baseline

No baseline yet. Compare against violet (#2420, lr=3e-4) and gilbert (#2421, lr=1e-3).

- **Primary metric:** `val_primary/surface_mse` (lower is better)
- **Also report:** `val_primary/volume_mse`, per-channel MSEs for `Ux`, `Uy`, `p`, `nut`

## Results

_To be filled in by the student._

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT senku:

## Results

### Run Summary
- **W&B run:** [`c6ny2kjc`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/c6ny2kjc)
- **Epochs completed:** 6 (30-min timeout)
- **Cosine T-max:** 50 (adjusted from 150 to match epoch cap)
- **Peak memory:** Not tracked in train.py

### Command
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --epochs 150 --lr 5e-4 \
  --optimizer lion --batch-size 2 --model senpai_transolver \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --model-slices 96 --model-mlp-ratio 4 --model-dropout 0.0 \
  --cosine-t-max 50 --use-ema --ema-decay 0.999 --ema-start-step 50 \
  --use-lookahead --surface-refine --surface-refine-hidden 128 \
  --surface-refine-layers 2 --weight-decay 1e-4 \
  --agent senku --wandb-name "senku/airfrans-lr-5e4" \
  --wandb-group "radford-airfrans-round1"
```

### Primary Metric: `val_primary/surface_mse`

| Epoch | val_primary/surface_mse | full_val/volume_mse | LR |
|-------|------------------------|--------------------|----|
| 1 | 0.5952 | 0.9157 | 4.77e-5 |
| 2 | 0.5989 | 0.9039 | 3.27e-4 |
| 3 | 0.6535 | 0.8713 | 3.27e-4 |
| **4** | **0.4151** | **0.8926** | 4.77e-5 |
| 5 | 0.5757 | 1.0695 | 5.00e-4 |
| 6 | 0.5940 | 1.3153 | 4.77e-5 |

**Best val surface MSE: 0.4151** (epoch 4)

### Final Test Metrics (EMA model, epoch 6)

| Metric | Value |
|--------|-------|
| **test_primary/surface_mse** | **0.5863** |
| full_test/volume_mse | 1.3354 |
| surface_mse_Ux | 0.00717 |
| surface_mse_Uy | 0.00022 |
| surface_mse_nut | 0.00034 |
| surface_mse_p | 2.3374 |
| volume_mse_Ux | 1.1376 |
| volume_mse_Uy | 1.9797 |
| volume_mse_nut | 1.0590 |
| volume_mse_p | 1.1654 |

### Analysis

**What happened:** The run completed 6 epochs within the 30-minute timeout. The best validation surface MSE of 0.4151 occurred at epoch 4 when the LR was at a cosine trough (4.77e-5). When the LR peaked at 5e-4 (epoch 5), both surface and volume metrics degraded noticeably — surface MSE jumped to 0.5757 and volume MSE spiked to 1.0695. This suggests lr=5e-4 may be too aggressive for this architecture/optimizer combination, as the model struggled to maintain performance at the peak learning rate.

The pressure channel (`p`) dominates the surface MSE (~2.34 at test), which is consistent across all epochs. Velocity and turbulence channels are well-behaved (Ux ~0.007, Uy ~0.0002, nut ~0.0003).

**Note:** Only 6 epochs ran due to the 30-min timeout, so these are early-training results. The cosine schedule was adjusted to t_max=50 (matching the epoch cap) rather than the PR's original 150. A longer run would be needed for definitive conclusions, but the early signal at peak LR is informative.

### Suggested follow-ups
- Compare these early-epoch curves with violet (#2420, lr=3e-4) and gilbert (#2421, lr=1e-3) at the same epoch count for a fair LR sensitivity comparison.
- The degradation at peak LR=5e-4 hints that the sweet spot may be closer to 3e-4. A run at lr=4e-4 could narrow the bracket.
- Consider adding peak GPU memory logging to train.py for future experiments.

**Comment by morganmcg1** [comment]
ADVISOR: Use this PR to fill in the mid-LR AirfRANS response curve, not just to test exactly `5e-4` once.

Launch at least these 4 variants in parallel:
- `v0`: `--lr 4e-4`
- `v1`: `--lr 5e-4`
- `v2`: `--lr 6e-4`
- `v3`: `--lr 7e-4`

Keep the optimizer and model fixed otherwise. Use one GPU per trial via `CUDA_VISIBLE_DEVICES`, keep the same W&B group, and suffix the names. This PR should return a local LR bracket around the current default regime, not a single intermediate run.

**Comment by morganmcg1** [comment]
ADVISOR: Clarification on the effective run budget for this `radford` sweep.

The assignment text may show `--epochs 150`, but the live senpai harness for this run is enforcing:
- `SENPAI_MAX_EPOCHS=50`
- `SENPAI_TIMEOUT_MINUTES=30`

So in practice, your run stops at 50 epochs or 30 minutes, whichever comes first. Treat this as the true budget when deciding how many parallel variants to launch and when interpreting results.

Please report for each variant:
- epochs actually completed
- whether it hit the 30-minute timeout or the 50-epoch cap
- the best validation metric reached within that enforced budget

Optimize for best performance under this actual capped regime, not for hypothetical longer training beyond the harness limits.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2422 because Lion lr=5e-4 (0.594) is worse than both the Lion baseline and Lion lr=1e-3. Senku's cosine T_max=50 adjustment was the right instinct (helps the cosine schedule match the actual epoch budget), but the fundamental issue is optimizer: AdamW at the same lr=5e-4 achieved 0.331. Round 2 will explore cosine schedule tuning under AdamW..
